src/main/java/com/scylladb/cdc/debezium/connector/transforms/ScyllaExtractNewRecordState.java

```diff
@@ -51,6 +57,48 @@ public void close() {
         schemaUpdateCache = null;
     }
 
+    // createSchemaKey computes a unique schema according to the payload values
+    // This ensures different values are stored with different keys in the LRU cache.
+    private String createSchemaKey(Struct value) {
+       Schema schema = value.schema();
+       StringBuilder key = new StringBuilder();
+       key.append(schema.name()).append(":");
+
+       for (Field field : schema.fields()) {
+            key.append(field.name()).append("-");
+
+            Object fieldValue = value.get(field);
+            boolean isNull = fieldValue == null;
+            key.append(isNull ? "N" : "P");
+
+            // Struct types
+            if (!isNull && field.schema().type() == Type.STRUCT) {
+                key.append("-");
+                Struct structValue = (Struct) fieldValue;
+                for (Field nestedField : structValue.schema().fields()) {
+                    if (structValue.get(nestedField) != null) {
+                        key.append(nestedField.name()).append(";");
+                    }
+                }
+            }
+       }
+
+       return key.toString();
+    }
+
+    // isDroppableField drops null Struct values within a Field. It does NOT drops empty Structs representing an actual delta mutation.
+    // This allow subscribers to upsert deltas and converge to a stable result. For example, the following Type/Values will return:
+    // False (don't drop) -- Type: STRUCT Value: Struct{}
+    // True (drop)        -- Type: STRUCT Value: null
+    private boolean isDroppableField(Field field, Struct value) {
+        if (field.schema().type() == Type.STRUCT
```

Comment:
Shouldn't you consider here if field is optional or not.
If it is not optional you can't just drop it, it will make data invalid.

Reply:
Probably it should. Although, if one produces a required field with a null value, then that's already a problem on its own. lol

Interestingly enough, I wasn't able to find a single message generated under this SMT with `optional: false`. I think because this SMT was originally made for RDBMs the field should be set somewhere upstream depending on the table schema (ie: a defined primary key when parsing the binlog?)